docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+
+##### Include Device Names
+
+Optional list of Device Names to filter for
+
+##### Exclude Device Profile Names
+
+Optional list of Device Profile Names to filter out
+
+##### Exclude Device Names
+
+Optional list of Device Names to filter out
+
+### Replay endpoint
+
+This `POST` API will start replaying the recorded data as specified in the [request Data Transfer Object (DTO)](#replay-request-dto) defined below. The request handler will validate the DTO and that the appropriate Device Profiles and Devices from the data exist. It will then start an async Go function to handle the replay so the request doesn't timeout on long replays. 
+
+The replay async Go function will use the [Background Publishing](https://docs.edgexfoundry.org/2.3/microservices/application/AdvancedTopics/#background-publishing) capability to send the recorded Events to the EdgeX MessageBus using the same publish topic scheme used by Device Services, which is `edgex/events/device/<device-profile-name>/<device-name>/<source-name>`. The App SDK has the  [Publish Topic Placeholders](https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#publish-topic-placeholders) capability built-in to facilitate this. The data for these topics is available from the Event DTO. The timestamps in the Events and Readings published will be set to the current date/time. This requires a copy be made of the Event/Readings as they are published in order to not corrupt the original data.
+
+Once the first event is published the replay function will calculate the wait time to use before sending the next Event from the recorded data. This will be based on the time difference from the original timestamp of the previous event published and the timestamp of the next event multiplied by the inverse of the `Replay Rate` specified in the request DTO. 
+
+!!! example - "Examples - Replay Rate wait time calculation"
+    Delta time between original Events is 800ms<br/>
+    Replay rate is 2.0 (100% faster) making wait time 400ms (800ms * (1 / 2.0))<br/>
+    Replay rate is 0.5 (100% slower) making wait time 1600ms (800ms * (1 / 0.5))
+
+The replay function will repeat publishing the recorded data per the `Repeat Count` in from the DTO. 
+
+#### Replay Request DTO
+
+##### Replay Rate
+
+Required rate at which to replay the data compared to the rate the data was recorded. Float value greater than 0 where 1 is the same rate, less than 1 is slower rate and greater than 1 is faster rate than the rate the data was recorded. 
+
+##### Repeat Count
+
+Optional count of number of times to repeat the replay. Default is 1 if not specified or value is 0.
+
+### Download endpoint (Export)
+
+This `GET` API will request that the previously recorded data be exported as a file download. It will accept an optional query parameter to specify compression (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.
```

Comment:
In HTTP requests, the compression is usually requested using the [Accept-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) header.

Reply:
> Having a web service adds several complications for no obvious gain.

I disagree. Having a service makes it more accessible. If you prefer the CLI experience you can use curl commands or as I prefer POSTMan or the UI can have a tab to control the service.

Reply:
[Accept-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) tells the client what the service supports, not what the client is requesting.
IMO, the query parameter is appropriate.

Reply:
No, that is not correct.
> The Accept-Encoding request HTTP header indicates the content encoding (usually a compression algorithm) that the client can understand. 

> IMO, the query parameter is appropriate.
If you want to not follow the standard practices and make your own, yes it is appropriate.

Reply:
It is more complicated than that. The server uses [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) to select one of the proposals and informs the client of that choice with the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) response header. 

Yes, I am fine with "none standard" if that is what you feel it is. ;-)